### PR TITLE
implemented ITilesProvider in SqlDataProvider

### PR DESCRIPTION
**Pull request by danlla**

### CODE REVIEW

File: OgcApi.Net/Crs/CoordinateConverter.cs [hunk 1]
```diff
@@ -29,5 +29,13 @@ public static double TileYToLat(int y, int z)
             var n = Math.PI - 2.0 * Math.PI * y / (1 << z);
             return 180.0 / Math.PI * Math.Atan(0.5 * (Math.Exp(n) - Math.Exp(-n)));
         }
+        public static double[] TileBounds(int x, int y, int z)
```
Comment (kolyanch) [inline OgcApi.Net/Crs/CoordinateConverter.cs:32]: Add an empty line before this one.
Consider returning Envelope object instead of an array of double.

**Review by kolyanch — CHANGES_REQUESTED** [review]

**Review by kolyanch — APPROVED** [review]

**Review by kolyanch — APPROVED** [review]

File: OgcApi.SqlServer.Tests/SqlServerFacts.cs [hunk 2]
```diff
@@ -578,5 +578,45 @@ public void DeleteNotExistingFeature()
...
+        public async void GetTileIncorrectTileCol()
+        {
+            var tile = await TestProviders.GetDefaultProvider().GetTileAsync("Polygons", 8, 170, 82);
+            Assert.Null(tile);
+        }
+
+        [Fact]
+        public void GetLimitsUnknownCollection()
+        {
+            Assert.Throws<ArgumentException>(() => TestProviders.GetDefaultProvider().GetLimits("test"));
```
Comment (bam-samis) [inline OgcApi.SqlServer.Tests/SqlServerFacts.cs:619]: The test doesn't work. Logically, if you ask for TileMatrixLimits for a collection that doesn't exist in the provider's settings, you should generate an exception. See how this is done in MBTiles.

File: OgcApi.SqlServer.Tests/SqlServerFacts.cs [hunk 3]
```diff
@@ -578,5 +578,45 @@ public void DeleteNotExistingFeature()
...
+            var tile = await TestProviders.GetDefaultProvider().GetTileAsync("Polygons", 8, 162, 90);
+            Assert.Null(tile);
+        }
+
+        [Fact]
+        public async void GetTileIncorrectTileCol()
+        {
+            var tile = await TestProviders.GetDefaultProvider().GetTileAsync("Polygons", 8, 170, 82);
+            Assert.Null(tile);
+        }
```
Comment (bam-samis) [inline OgcApi.SqlServer.Tests/SqlServerFacts.cs:614]: These tests make no sense since you have a predefined TileMatrixSet. Delete. Instead, add a test to check that an empty tile (without objects) is returned when the corresponding x, y, and z are specified.

File: OgcApi.Postgis.Tests/PostGisFacts.cs [hunk 4]
```diff
@@ -576,5 +576,45 @@ public void DeleteNotExistingFeature()
             var provider = TestProviders.GetDefaultProvider();
             Assert.Throws<ArgumentException>(() => provider.DeleteFeature("PolygonsForInsert", "200"));
         }
+
```
Comment (bam-samis) [inline OgcApi.Postgis.Tests/PostGisFacts.cs:579]: The remarks are similar to SqlServerFacts.

File: OgcApi.Net/DataProviders/SqlDataProvider.cs [hunk 5]
```diff
@@ -493,5 +500,45 @@ public void DeleteFeature(string collectionId, string featureId, string apiKey =
...
 
         protected abstract Geometry ReadGeometry(DbDataReader dataReader, int ordinal, SqlFeaturesSourceOptions collectionSourceOptions);
+
+        public Task<byte[]> GetTileAsync(string collectionId, int tileMatrix, int tileRow, int tileCol, string apiKey = null)
+        {
+            var bbox = CoordinateConverter.TileBounds(tileRow, tileCol, tileMatrix);
+            var features = GetFeatures(collectionId, bbox: bbox);
+
+            var tileDefinition = new NetTopologySuite.IO.VectorTiles.Tiles.Tile(tileMatrix, tileRow, tileCol);
+            var vectorTile = new VectorTile { TileId = tileDefinition.Id };
```
Comment (bam-samis) [inline OgcApi.Net/DataProviders/SqlDataProvider.cs:510]: Tile ID is not needed, replace with
var vectorTile = new VectorTile();

File: OgcApi.Net/DataProviders/SqlDataProvider.cs [hunk 6]
```diff
@@ -493,5 +500,45 @@ public void DeleteFeature(string collectionId, string featureId, string apiKey =
...
+            {
+                layer.Features.Add(feature);
+            }
+
+            vectorTile.Layers.Add(layer);
+
+            using var memoryStream = new MemoryStream();
+
+            vectorTile.Write(memoryStream);
+
```
Comment (bam-samis) [inline OgcApi.Net/DataProviders/SqlDataProvider.cs:523]: Add memoryStream.Seek(0, SeekOrigin.Begin);

**Review by bam-samis — CHANGES_REQUESTED** [review]
Remove unnecessary tests. Add a test to return an empty tile. Minor code changes.

File: OgcApi.Net/DataProviders/SqlDataProvider.cs [hunk 7]
```diff
@@ -493,5 +500,52 @@ public void DeleteFeature(string collectionId, string featureId, string apiKey =
...
+            var collectionOptions = (CollectionOptions)CollectionsOptions.GetSourceById(collectionId);
+            if (collectionOptions == null)
+            {
+                Logger.LogTrace(
+                    $"The source collection with ID = {collectionId} was not found in the provided options");
+                throw new ArgumentException($"The source collection with ID = {collectionId} does not exists");
+            }
+
+            var result = new List<TileMatrixLimits>();
+            for (var i = 0; i <= 22; ++i)
```
Comment (bam-samis) [inline OgcApi.Net/DataProviders/SqlDataProvider.cs:537]: In the for loop, replace prefix increment with postfix increment. In this context it makes no sense to use prefix increment.

File: OgcApi.SqlServer.Tests/SqlServerFacts.cs [hunk 8]
```diff
@@ -578,5 +580,34 @@ public void DeleteNotExistingFeature()
...
+        [Fact]
+        public async void GetTileEmptyTile()
+        {
+            var rawTile = await TestProviders.GetDefaultProvider().GetTileAsync("LineStrings", 8, 1, 250);
+            using var memoryStream = new MemoryStream(rawTile);
+            memoryStream.Seek(0, SeekOrigin.Begin);
+            var tile = new MapboxTileReader().Read(memoryStream, new NetTopologySuite.IO.VectorTiles.Tiles.Tile(8,1,250));
+            Assert.True(tile.IsEmpty);
+        }
+
```
Comment (bam-samis) [inline OgcApi.SqlServer.Tests/SqlServerFacts.cs:606]: Add a GetLimits test for the correct collection name in which you check that the result matches expectations.

**Review by bam-samis — CHANGES_REQUESTED** [review]

File: OgcApi.Postgis.Tests/PostGisFacts.cs [hunk 9]
```diff
@@ -595,10 +596,17 @@ public void GetTileUnknownCollection()
         [Fact]
         public async void GetTileEmptyTile()
         {
-            var rawTile = await TestProviders.GetDefaultProvider().GetTileAsync("LineStrings", 8, 1, 250);
+            var rawTile = await TestProviders.GetDefaultProvider().GetTileAsync("Polygons", 8, 1, 250);
             using var memoryStream = new MemoryStream(rawTile);
             memoryStream.Seek(0, SeekOrigin.Begin);
-            var tile = new MapboxTileReader().Read(memoryStream, new NetTopologySuite.IO.VectorTiles.Tiles.Tile(8, 1, 250));
+
+            using var decompressor = new GZipStream(memoryStream, CompressionMode.Decompress);
```
Comment (kolyanch) [inline OgcApi.Postgis.Tests/PostGisFacts.cs:603]: await using

File: OgcApi.Postgis.Tests/PostGisFacts.cs [hunk 10]
```diff
@@ -595,10 +596,17 @@ public void GetTileUnknownCollection()
...
+
+            using var decompressedStream = new MemoryStream();
+            decompressor.CopyTo(decompressedStream);
```
Comment (kolyanch) [inline OgcApi.Postgis.Tests/PostGisFacts.cs:606]: Use CopyToAsync

File: OgcApi.SqlServer.Tests/SqlServerFacts.cs [hunk 11]
```diff
@@ -597,10 +598,17 @@ public void GetTileUnknownCollection()
         [Fact]
         public async void GetTileEmptyTile()
         {
-            var rawTile = await TestProviders.GetDefaultProvider().GetTileAsync("LineStrings", 8, 1, 250);
+            var rawTile = await TestProviders.GetDefaultProvider().GetTileAsync("Polygons", 8, 1, 250);
             using var memoryStream = new MemoryStream(rawTile);
             memoryStream.Seek(0, SeekOrigin.Begin);
-            var tile = new MapboxTileReader().Read(memoryStream, new NetTopologySuite.IO.VectorTiles.Tiles.Tile(8, 1, 250));
+
+            using var decompressor = new GZipStream(memoryStream, CompressionMode.Decompress);
```
Comment (kolyanch) [inline OgcApi.SqlServer.Tests/SqlServerFacts.cs:605]: await using

File: OgcApi.SqlServer.Tests/SqlServerFacts.cs [hunk 12]
```diff
@@ -597,10 +598,17 @@ public void GetTileUnknownCollection()
...
+
+            using var decompressedStream = new MemoryStream();
+            decompressor.CopyTo(decompressedStream);
```
Comment (kolyanch) [inline OgcApi.SqlServer.Tests/SqlServerFacts.cs:608]: Use CopyToAsync

**Review by kolyanch — CHANGES_REQUESTED** [review]
